fe/fe-core/src/main/java/com/starrocks/meta/lock/LockManager.java

```diff
@@ -0,0 +1,521 @@
+                } else if (lock instanceof LightWeightLock) {
+                    List<LockHolder> owners = new ArrayList<>(lock.getOwners());
+                    if (owners.size() > 0) {
+                        /* Lock is already held by someone else so mutate. */
+                        lock = new MultiUserLock(owners.get(0));
+                        lockTable.put(rid, lock);
+                    }
+                }
+
+                LockGrantType lockGrantType = lock.lock(locker, lockType);
+                if (lockGrantType == LockGrantType.NEW || lockGrantType == LockGrantType.EXISTING) {
+                    return;
+                }
+            }
+
+            startTime = System.currentTimeMillis();
+            locker.setWaitingFor(rid, lockType);
+
+            /*
+             * Because deadlock detection also requires a significant cost, but at the first moment
+             * when a lock cannot be obtained, we cannot determine whether it is because the required
+             * lock is being used normally or if a deadlock has occurred.
+             * Therefore, based on the configuration parameter `dead_lock_detection_delay_time_ms`
+             * is used to control the waiting time before deadlock detection.
+             * If a lock is obtained during this period, there is no need to perform deadlock detection.
+             * Avoid frequent and unnecessary deadlock detection due to lock contention
+             */
+            long deadLockDetectionDelayTimeMs = Config.dead_lock_detection_delay_time_ms;
+            if (Config.dead_lock_detection && deadLockDetectionDelayTimeMs > 0) {
+                if (timeout != 0) {
+                    deadLockDetectionDelayTimeMs = Math.min(deadLockDetectionDelayTimeMs, timeRemain(timeout, startTime));
+                }
+
+                try {
+                    locker.wait(Math.max(1, deadLockDetectionDelayTimeMs));
+                } catch (InterruptedException ie) {
+                    removeFromWaiterList(rid, locker, lockType);
+                    throw new LockInterruptException(ie);
+                }
+
+                if (isOwner(rid, locker, lockType)) {
+                    locker.clearWaitingFor();
+                    return;
+                }
+            }
+
+            /*
+             * After waiting, not acquire lock and entered the waiting period, with deadlock detection enabled
+             */
+        }
+
+        while (true) {
+            Locker victim = null;
+            synchronized (locker) {
+                while (true) {
+                    if (isOwner(rid, locker, lockType)) {
+                        break;
+                    }
+
+                    if (Config.dead_lock_detection) {
+                        victim = checkAndHandleDeadLock(rid, locker, lockType);
+                    }
+                    if (victim != null) {
+                        /* Locker owns the lock and no deadlock was detected. */
```

Comment:
wrong comment? there's dead lock, but just the current locker is not chosen as the victim?